SharedRootEventSource.cc

```diff
@@ -169,33 +169,38 @@ void SharedRootEventSource::readEventAsync(unsigned int iLane, long iEventIndex,
   queue_.push(*iTask.group(), [iLane, optTask = std::move(iTask), this, iEventIndex]() mutable {
       auto start = std::chrono::high_resolution_clock::now();
       if(iEventIndex < eventsTree_->GetEntries()) {
-        std::vector<uint32_t> buffer;
-        auto pBuffer = &buffer;
+        std::pair<std::vector<uint32_t>, std::vector<char>> offsetsAndBuffer;
+        auto pBuffer = &offsetsAndBuffer;
         eventsBranch_->SetAddress(&pBuffer);
+
         idBranch_->SetAddress(&this->laneInfos_[iLane].eventID_);
         eventsTree_->GetEntry(iEventIndex);
         {
-          auto const& id = this->laneInfos_[iLane].eventID_;
+          //auto const& id = this->laneInfos_[iLane].eventID_;
+          //std::cout <<"event entry "<<iEventIndex<<std::endl;
           //std::cout <<"ID "<<id.run<<" "<<id.lumi<<" "<<id.event<<std::endl;
           //std::cout <<"buffer size "<<buffer.size()<<std::endl;
+          //std::cout <<"offset size "<<offsets.size()<<std::endl;
           //for(auto b: buffer) {
           //  std::cout <<"  "<<b<<std::endl;
           //}
         }
 
         auto group = optTask.group();
-        group->run([this, buffer=std::move(buffer), task = optTask.releaseToTaskHolder(), iLane]() {
+        group->run([this, offsetsAndBuffer=std::move(offsetsAndBuffer), task = optTask.releaseToTaskHolder(), iLane]() {
             auto& laneInfo = this->laneInfos_[iLane];
 
             auto start = std::chrono::high_resolution_clock::now();
-            std::vector<uint32_t> uBuffer = pds::uncompressEventBuffer(this->compression_, buffer);
-            //std::cout <<"uncompressed buffer size "<<uBuffer.size() <<std::endl;
+            std::vector<char> uBuffer = pds::uncompressBuffer(this->compression_, offsetsAndBuffer.second, offsetsAndBuffer.first.back());
+            std::cout <<"uncompressed buffer size "<<uBuffer.size() <<std::endl;
             laneInfo.decompressTime_ += 
               std::chrono::duration_cast<decltype(laneInfo.decompressTime_)>(std::chrono::high_resolution_clock::now() - start);
             
             start = std::chrono::high_resolution_clock::now();
             //uBuffer.pop_back();
-            pds::deserializeDataProducts(uBuffer.begin(), uBuffer.end(), laneInfo.dataProducts_, laneInfo.deserializers_);
+            pds::deserializeDataProducts(&(*uBuffer.begin()), &(*uBuffer.end()), 
```

Comment:
NOTE: in this implementation we deserialize all the data all at once. In a future version I hope to see if there is a benefit to allowing that to happen concurrently when the data is requested. The existence of the separate _offset_ information would make that trivial.